WT-15773 Fix unit test on MacOS

ablekhman opened the pull request:
- If `ENABLE_PALITE=ON` and compiler does not support PALite build, then make it a cmake error.
- Run `unit-test-macos` with PALM for now, until newer compiler is available for MacOS builds.

ablekhman committed:
WT-15773 Fix unit test on MacOS
- If `ENABLE_PALITE=ON` and compiler does not suport PALite build, then make it a cmake error.
- Run `unit-test-macos` with APLM for now, until newer compiler is available for MacOS builds.

ablekhman committed:
WT-15773 Fix unit test on MacOS
- If `ENABLE_PALITE=ON` and compiler does not suport PALite build, then make it a cmake error.
- Run `unit-test-macos` with APLM for now, until newer compiler is available for MacOS builds.
- Disable PALite build for older compilers

ablekhman committed:
WT-15773 Fix unit test on MacOS

- If `ENABLE_PALITE=ON` and compiler does not suport PALite build, then make it a cmake error.
- Run `unit-test-macos` with APLM for now, until newer compiler is available for MacOS builds.
- Disable PALite build for older compilers

ablekhman committed:
WT-15773 Fix unit test on MacOS

- If `ENABLE_PALITE=ON` and compiler does not suport PALite build, then make it a cmake error.
- Run `unit-test-macos` with APLM for now, until newer compiler is available for MacOS builds.
- Disable PALite build for older compilers and doc builds

jiechenbo commented on test/evergreen.yml:
```diff
@@ -320,6 +321,8 @@ functions:
 
           if [ "${build_variant|}" = "macos-14-arm64" ]; then
             DEFINED_EVERGREEN_CONFIG_FLAGS="$DEFINED_EVERGREEN_CONFIG_FLAGS -DPython3_EXECUTABLE=$SYSPY -DPython3_LIBRARY=$SYSPYLIB -DPython3_INCLUDE_DIR=$SYSPYINCDEF"
+            # FIXME-WT-15777 PALite currently does not build on macOS 14 (ARM64) because of outdated CLang.
```
I'd rather configuration this under macos-14-arm64 variant rather than here.

ablekhman replied:
I moved it to the `mac_test_template` section. So, all the flags are in one place.

jiechenbo commented on test/evergreen.yml:
```diff
@@ -555,8 +558,9 @@ functions:
-              $CMAKE -DCMAKE_C_FLAGS="-DMIGHT_NOT_RUN -Wno-error" -DENABLE_PYTHON=1 -DENABLE_STRICT=1 ../.
+              # Adding -DENABLE_PYTHON=ON -DENABLE_STRICT=ON as 6.0 does not default these like develop.
+              # Disable PALite build, as docs build does not need it and it can fail to build on some platforms.
+              $CMAKE -DCMAKE_C_FLAGS="-DMIGHT_NOT_RUN -Wno-error" -DENABLE_PYTHON=ON -DENABLE_STRICT=ON -DENABLE_PALITE=OFF ../.
```
ditto

ablekhman replied:
This build runs on other distros. It is not specific to macos.

jiechenbo replied:
This implies default is DENABLE_PALITE=ON, instead i only want to turn on PALITE=ON when we need to use disagg.

ablekhman replied:
Right now we cannot distinguish between disagg and non-disagg tests/builds. The `unit-test*` tasks run everything.
It's much bigger project to separate disagg and non-disagg builds. The practical solution is to disable PALite for certain builds where no tests expected to run.

jiechenbo commented on test/evergreen.yml:
```diff
@@ -1745,30 +1755,37 @@ tasks:
         vars:
           CLANG_C_VERSION: -DCLANG_C_VERSION=13
           CLANG_CXX_VERSION: -DCLANG_CXX_VERSION=13
+          ENABLE_PALITE: -DENABLE_PALITE=OFF
```
We should be doing all of this at the macos variant instead of tasks/functions.

ablekhman replied:
These compilation tasks with older compilers run on other distros. They are not related to macos, just found them recently.

Context: I made it a cmake error when user requires `ENABLE_PALITE=ON`, but compiler cannot build it. In the past PALite was simply skipped with cmake warning. Making it an error is more user friendly. People will see the error early, instead of being surprised by failing tests later.
That's why other distros need to exclude PALite from the build, if running tests is not needed (doc builds, etc).

jiechenbo replied:
Hmm I'm not a fan of this, ideally if we are testing on older compiler versions we don't need to mindful whether PALite will cause us a compilation issue. Especially since PAlite is not needed for classic WiredTiger.

ablekhman replied:
The problem with quietly skipping PALite build, is that people discover it much later when their Python tests start failing with strange errors. At least, on Evergreen we do it once and it's not for human consumption.

jiechenbo replied:
I would like a second opinion to make sure we are okay here. cc: @korteland

korteland replied:
Per the offline conversation, I made WT-15805

ablekhman committed:
WT-15773 Fix unit test on MacOS

- If `ENABLE_PALITE=ON` and compiler does not suport PALite build, then make it a cmake error.
- Run `unit-test-macos` with APLM for now, until newer compiler is available for MacOS builds.
- Disable PALite build for older compilers and doc builds

jiechenbo reviewed (approved):
LGTM, except for one comment please get it resolved before merging PR
